### PR TITLE
Command `tag`: improve error and status messages

### LINKED ISSUE #225: tag subcommand fails with unhelpful error message when the tag exists
Running `dune-release tag` fails if the tag already exists:
```
$ dune-release tag
[-] Extracting tag from first entry in CHANGES.md
[-] Using tag "v0.2.0+test"
[?] Create git tag v0.2.0+test for HEAD? [Y/n]

dune-release: [ERROR] The following command exited with code 128:
                      git --git-dir .git tag -a -m Distribution v0.2.0+test
                      v0.2.0+test HEAD
```

This isn't very helpful. I think that to the very least we should:
1. clean up the error logs in those cases, I'm really not so fond of `Logs` default error reporting.
2. Show external commands stderr/stdout when they fail

In this particular case we could go even further and check whether the tag exists before tagging.If it exists and points to the current `HEAD`, we could successfully exit, informing the user that the tag already exists and points to the right place. If it points to somewhere else, we could fail and inform the user why we did.

### PR DESCRIPTION
**Pull request by pitag-ha**
Fixes #225.

### CODE REVIEW

File: lib/vcs.ml [hunk 1]
```diff
@@ -400,6 +400,14 @@ let tag_exists ~dry_run r tag =
   | (`Git, _, _) as r -> git_tag_exists r ~dry_run tag
   | `Hg, _, _ -> failwith "TODO"
 
+let tag_points_to_commit r tag commit_ish =
```
Comment (NathanReb) [inline lib/vcs.ml:403]: I'm not sure this function is required. What would you think about extracting a function:
```ocaml
(** [tag_points_to t tag] returns the commit hash [tag] points to or [None] if it doesn't exist *)
tag_points_to : dry_run: bool -> t -> string -> string option
```
and implement `tag_exists` on top of it. We can then clean the tag  command code a bit so that it starts by getting the `commit_ish` actual commit. Note that this we should fail if it doesn't exist anyway so might as well do it early. It would then get the commit for the tag and behave accordingly.

File: bin/tag.ml [hunk 2]
```diff
@@ -10,22 +10,54 @@ open Dune_release
...
   Vcs.get () >>= fun repo ->
+  let tag_exists = Vcs.tag_exists ~dry_run:false repo tag in
   match delete with
   | true ->
-      Prompt.confirm_or_abort ~yes ~question:(fun l ->
-          l "Delete tag %a?" Text.Pp.version tag)
-      >>= fun () ->
-      Vcs.delete_tag ~dry_run repo tag >>| fun () ->
-      App_log.success (fun m -> m "Deleted tag %a" Text.Pp.version tag)
+      if tag_exists then
```
Comment (NathanReb) [inline bin/tag.ml:16]: Following my suggestion regarding `tag_points_to`, I think this whole function could be simplified a bit, maybe as a single pattern matching over `delete`, `commit_ish` and the optional commit the tag points to.

It is a bit convoluted at the moment and makes it hard to follow. What do you think?

File: bin/tag.ml [hunk 3]
```diff
@@ -10,22 +10,54 @@ open Dune_release
...
+                   (default: HEAD). Do you want to delete it anyways?"
+                  Fmt.(styled `Red string)
+                  "Warning:" Text.Pp.version tag)
+        in
+        question >>= fun () ->
+        Vcs.delete_tag ~dry_run repo tag >>| fun () ->
+        App_log.success (fun m -> m "Deleted tag %a" Text.Pp.version tag)
+      else
+        Ok
+          (App_log.success (fun apply_log_l ->
```
Comment (NathanReb) [inline bin/tag.ml:36]: I'm wondering whether this should be a success or status since nothing actually happened. What do you think?

File: lib/prompt.ml [hunk 4]
```diff
@@ -1,22 +1,33 @@
...
   App_log.question (fun l ->
-      f (fun ?header ?tags fmt -> l ?header ?tags (fmt ^^ " [Y/n]")))
+      f (fun ?header ?tags fmt ->
+          l ?header ?tags (fmt ^^ Scanf.format_from_string options "")))
 
-let confirm ~question ~yes =
-  let rec loop () =
-    ask question;
+let rec loop ~question ~default_answer ~yes =
+  ask question ~default_answer;
```
Comment (NathanReb) [inline lib/prompt.ml:10]: As mentioned below, we shouldn't "ask" if `--yes` was passed and just go straight to the point so this should be moved to the else branch I think.

File: bin/tag.ml (hunk 3 above)
Comment (pitag-ha) [reply bin/tag.ml:36]: Yes, status definitely seems more appropriate here. I've also replaced `success` by `status` in the case of trying to create a tag that already exists providing a "right" commit-ish.

File: lib/vcs.ml (hunk 1 above)
Comment (pitag-ha) [reply lib/vcs.ml:403]: Yes, that structure you're suggesting is far nicer! I'm just not sure if I understand what you have in mind for `tag_points_to` and `tag_exists`. Could you have another look (once I push the changes)?

File: lib/prompt.ml [hunk 5]
```diff
@@ -1,22 +1,33 @@
-let ask f =
+type default = Yes | No
+
+let ask f ~default_answer =
+  let options = match default_answer with Yes -> " [Y/n]" | No -> " [y/N]" in
   App_log.question (fun l ->
-      f (fun ?header ?tags fmt -> l ?header ?tags (fmt ^^ " [Y/n]")))
+      f (fun ?header ?tags fmt ->
+          l ?header ?tags (fmt ^^ Scanf.format_from_string options "")))
```
[the remark this answers is not in the document]
Comment (pitag-ha) [reply lib/prompt.ml:7]: Cool! That works.

File: CHANGES.md [hunk 6]
```diff
@@ -7,9 +7,16 @@
 - Add command `delegate-info` to print information needed by external
   release scripts (#221, @pitag-ha)
 - Use Curly instead of Cmd to interact with github (#202, @gpetiot)
+- Add `Prompt.confirm_or_abort_neg`: Like `[Prompt.confirm_or_abort]`,
```
[the remark this answers is not in the document]
Comment (pitag-ha) [reply CHANGES.md:10]: That's good to know. Thanks!

File: bin/tag.ml (hunk 2 above)
Comment (pitag-ha) [reply bin/tag.ml:16]: Yes, that actually makes it far nicer! :) Thanks.

File: tests/bin/delegate_info/run.t [hunk 7]
```diff
@@ -23,6 +23,7 @@ We need to set up a git project for dune-release to work properly
     $ dune-release tag -y
     [-] Extracting tag from first entry in CHANGES.md
     [-] Using tag "0.1.0"
+    [?] Create git tag 0.1.0 for HEAD? [Y/n]
```
[the remark this answers is not in the document]
Comment (pitag-ha) [reply tests/bin/delegate_info/run.t:26]: Ok, the old behaviour is back :)

File: lib/prompt.ml (hunk 4 above)
Comment (pitag-ha) [reply lib/prompt.ml:10]: Yeah, asking a question when auto-confirming does feel a bit off... The reason why I did that was to be able to E2E test the different behaviour of `tag -d` when providing a commit-ish, that corresponds to the provided tag, or providing a "wrong" commit-ish (the difference lies in the question that asks for confirmation and in the default answer to that question). Do you think I should test that part of `tag` by adding a (general) flag `--answer-default` or should I simply not test that part?
